File: inference-v3/src/magnitude_engine/models/qwen35/formats/gguf.py
"""Read a GGUF directory into Qwen's architecture-owned weight roles.

This is the one place the Qwen architecture and the GGUF container meet. A new
container for this model is a new file beside this one.
"""

from pydantic import TypeAdapter

import magnitensor as mt
from magnitude_engine.models.qwen35.description import (
    AttentionWeights,
    BlockWeights,
    DenseDescription,
    DenseFeedForwardWeights,
    ExpertGeometry,
    Geometry,
    MixerKind,
    RecurrentWeights,
    RoutedFeedForwardWeights,
)
from magnitude_engine.weights.descriptor import WeightDescriptor
from magnitude_engine.weights.formats.gguf import Directory, GGUFFormat
from magnitude_engine.weights.identity import ArtifactIdentity


def describe(format: GGUFFormat) -> DenseDescription:
    architecture = format.directory.value("general.architecture")
    if architecture == "qwen35":
        return inspect_dense(format.directory, format.identity)
    if architecture == "qwen35moe":
        return inspect_moe(format.directory, format.identity)
    raise ValueError(f"unsupported Qwen GGUF architecture {architecture!r}")


def inspect_dense(directory: Directory, identity: ArtifactIdentity) -> DenseDescription:
    return _inspect(directory, identity, "qwen35", routed=False)


def inspect_moe(directory: Directory, identity: ArtifactIdentity) -> DenseDescription:
    return _inspect(directory, identity, "qwen35moe", routed=True)


def _inspect(
    directory: Directory,
    identity: ArtifactIdentity,
    architecture: str,
    *,
    routed: bool,
) -> DenseDescription:
    if directory.value("general.architecture") != architecture:
        raise ValueError(f"Qwen binding requires the {architecture} GGUF architecture")
    metadata = {item.name: item.value for item in directory.metadata}
    prefix_key = architecture + "."
    if metadata.get(prefix_key + "rope.scaling.type", "none") != "none":
        raise ValueError("scaled Qwen rotary definitions are not yet qualified")

    def integer(key: str) -> int:
        value = directory.value(prefix_key + key)
        if type(value) is not int or value <= 0:
            raise ValueError(f"{prefix_key}{key} must be a positive integer")
        return value

    def number(key: str) -> float:
        value = directory.value(prefix_key + key)
        if isinstance(value, bool) or not isinstance(value, (int, float)):
            raise ValueError(f"{prefix_key}{key} must be numeric")
        return float(value)

    # A container may bundle the speculative multi-token head as trailing
    # blocks. Those are not model layers: this program does not run them, and
    # counting them would shift every layer's mixer kind.
    speculative = metadata.get(prefix_key + "nextn_predict_layers", 0)
    if type(speculative) is not int or not 0 <= speculative < integer("block_count"):
        raise ValueError(f"{prefix_key}nextn_predict_layers must be below the block count")
    layer_count = integer("block_count") - speculative
    flags = metadata.get(prefix_key + "attention.recurrent_layers")
    if flags is None:
        interval = integer("full_attention_interval")
        kinds = tuple(
            MixerKind.ATTENTION if (i + 1) % interval == 0 else MixerKind.RECURRENT
            for i in range(layer_count)
        )
    else:
        if (
            not isinstance(flags, tuple)
            or len(flags) != layer_count
            or any(type(v) is not bool for v in flags)
        ):
            raise ValueError("invalid per-layer recurrent flags")
        kinds = tuple(MixerKind.RECURRENT if flag else MixerKind.ATTENTION for flag in flags)
    embedding = directory.tensor("token_embd.weight")
    if len(embedding.shape) != 2:
        raise ValueError("Qwen embedding must be a matrix")
    vocabulary = metadata.get("tokenizer.ggml.tokens")
    if vocabulary is not None and (
        not isinstance(vocabulary, tuple) or len(vocabulary) != embedding.shape[0]
    ):
        raise ValueError("tokenizer vocabulary differs from the embedding table")
    sections = directory.value(prefix_key + "rope.dimension_sections")
    if (
        not isinstance(sections, tuple)
        or len(sections) != 4
        or any(type(n) is not int for n in sections)
    ):
        raise ValueError("Qwen rotary sections must contain four integers")
    expert_geometry = (
        ExpertGeometry(
            count=integer("expert_count"),
            selected=integer("expert_used_count"),
            intermediate=integer("expert_feed_forward_length"),
            shared_intermediate=integer("expert_shared_feed_forward_length"),
        )
        if routed
        else None
    )
    intermediate = (
        expert_geometry.intermediate
        if expert_geometry is not None
        else integer("feed_forward_length")
    )
    geometry = Geometry(
        activation_dtype=mt.DType.BF16,
        hidden=integer("embedding_length"),
        intermediate=intermediate,
        vocabulary=embedding.shape[0],
        context_limit=integer("context_length"),
        layers=kinds,
        attention_heads=integer("attention.head_count"),
        kv_heads=integer("attention.head_count_kv"),
        attention_width=integer("attention.key_length"),
        rotary_width=integer("rope.dimension_count"),
        rotary_base=number("rope.freq_base"),
        rotary_sections=TypeAdapter(tuple[int, int, int, int]).validate_python(
            sections, strict=True
        ),
        epsilon=number("attention.layer_norm_rms_epsilon"),
        convolution_width=integer("ssm.conv_kernel"),
        recurrent_key_heads=integer("ssm.group_count"),
        recurrent_value_heads=integer("ssm.time_step_rank"),
        recurrent_width=integer("ssm.state_size"),
        experts=expert_geometry,
    )
    if integer("attention.value_length") != geometry.attention_width:
        raise ValueError("Qwen attention requires equal key/value head widths")
    if integer("ssm.inner_size") != geometry.recurrent_value_heads * geometry.recurrent_width:
        raise ValueError("Qwen recurrent inner size differs from head geometry")
    consumed: set[str] = set()

    def tensor(name: str, shape: tuple[int, ...]) -> WeightDescriptor:
        value = directory.tensor(name)
        if value.shape != shape:
            raise ValueError(f"Qwen weight {name}: expected {shape}, received {value.shape}")
        consumed.add(name)
        return WeightDescriptor(name=name, shape=shape)

    g = geometry
    embedding = tensor(embedding.name, (g.vocabulary, g.hidden))
    output_norm = tensor("output_norm.weight", (g.hidden,))
    output_name = (
        "output.weight"
        if any(t.name == "output.weight" for t in directory.tensors)
        else embedding.name
    )
    output = tensor(output_name, (g.vocabulary, g.hidden))
    blocks = []
    for i, kind in enumerate(kinds):
        prefix = f"blk.{i}."
        if kind == MixerKind.ATTENTION:
            mixer = AttentionWeights(
                query_gate=tensor(
                    prefix + "attn_q.weight", (2 * g.attention_heads * g.attention_width, g.hidden)
                ),
                key=tensor(prefix + "attn_k.weight", (g.kv_heads * g.attention_width, g.hidden)),
                value=tensor(prefix + "attn_v.weight", (g.kv_heads * g.attention_width, g.hidden)),
                query_norm=tensor(prefix + "attn_q_norm.weight", (g.attention_width,)),
                key_norm=tensor(prefix + "attn_k_norm.weight", (g.attention_width,)),
                output=tensor(
                    prefix + "attn_output.weight", (g.hidden, g.attention_heads * g.attention_width)
                ),
            )
        else:
            mixer = RecurrentWeights(
                query_key_value=tensor(
                    prefix + "attn_qkv.weight", (g.recurrent_channels, g.hidden)
                ),
                gate=tensor(
                    prefix + "attn_gate.weight",
                    (g.recurrent_value_heads * g.recurrent_width, g.hidden),
                ),
                alpha=tensor(prefix + "ssm_alpha.weight", (g.recurrent_value_heads, g.hidden)),
                beta=tensor(prefix + "ssm_beta.weight", (g.recurrent_value_heads, g.hidden)),
                convolution=tensor(
                    prefix + "ssm_conv1d.weight", (g.recurrent_channels, g.convolution_width)
                ),
                decay=tensor(prefix + "ssm_a", (g.recurrent_value_heads,)),
                time_bias=tensor(prefix + "ssm_dt.bias", (g.recurrent_value_heads,)),
                norm=tensor(prefix + "ssm_norm.weight", (g.recurrent_width,)),
                output=tensor(
                    prefix + "ssm_out.weight",
                    (g.hidden, g.recurrent_value_heads * g.recurrent_width),
                ),
            )
        if expert_geometry is None:
            feedforward = DenseFeedForwardWeights(
                gate=tensor(prefix + "ffn_gate.weight", (g.intermediate, g.hidden)),
                up=tensor(prefix + "ffn_up.weight", (g.intermediate, g.hidden)),
                down=tensor(prefix + "ffn_down.weight", (g.hidden, g.intermediate)),
            )
        else:
            e = expert_geometry
            feedforward = RoutedFeedForwardWeights(
                router=tensor(prefix + "ffn_gate_inp.weight", (e.count, g.hidden)),
                shared_router=tensor(prefix + "ffn_gate_inp_shexp.weight", (g.hidden,)),
                expert_gate=tensor(
                    prefix + "ffn_gate_exps.weight", (e.count, e.intermediate, g.hidden)
                ),
                expert_up=tensor(
                    prefix + "ffn_up_exps.weight", (e.count, e.intermediate, g.hidden)
                ),
                expert_down=tensor(
                    prefix + "ffn_down_exps.weight", (e.count, g.hidden, e.intermediate)
                ),
                shared_gate=tensor(
                    prefix + "ffn_gate_shexp.weight", (e.shared_intermediate, g.hidden)
                ),
                shared_up=tensor(prefix + "ffn_up_shexp.weight", (e.shared_intermediate, g.hidden)),
                shared_down=tensor(
                    prefix + "ffn_down_shexp.weight", (g.hidden, e.shared_intermediate)
                ),
            )
        blocks.append(
            BlockWeights(
                input_norm=tensor(prefix + "attn_norm.weight", (g.hidden,)),
                mixer=mixer,
                feedforward_norm=tensor(prefix + "post_attention_norm.weight", (g.hidden,)),
                feedforward=feedforward,
            )
        )
    speculative_blocks = tuple(f"blk.{layer_count + i}." for i in range(speculative))
    unused = {
        tensor.name
        for tensor in directory.tensors
        if tensor.name not in consumed and not tensor.name.startswith(speculative_blocks or ("\0",))
    }
    if unused:
        raise ValueError(f"dense Qwen artifact contains unbound weight roles: {sorted(unused)}")
    return DenseDescription(
        artifact_identity=identity,
        geometry=g,
        embedding=embedding,
        output_norm=output_norm,
        output=output,
        blocks=tuple(blocks),
    )

File: inference-v4/engine/src/models/qwen35/gguf.rs
//! The boundary between Qwen architecture roles and the GGUF container.
use super::*;
use crate::weights::{
    descriptor::Transform,
    gguf::{Directory, Scalar, Value},
};
use std::collections::HashSet;
struct Metadata<'a> {
    directory: &'a Directory,
    prefix: String,
}
impl Metadata<'_> {
    fn value(&self, key: &str) -> Result<&Value, Error> {
        self.directory
            .value(&format!("{}{key}", self.prefix))
            .ok_or_else(|| invalid(format!("missing Qwen metadata {}{key}", self.prefix)))
    }
    fn optional(&self, key: &str) -> Option<&Value> {
        self.directory.value(&format!("{}{key}", self.prefix))
    }
    fn integer(&self, key: &str) -> Result<u64, Error> {
        self.value(key)?
            .unsigned()
            .filter(|n| *n > 0)
            .ok_or_else(|| invalid(format!("{}{key} must be a positive integer", self.prefix)))
    }
    fn number(&self, key: &str) -> Result<f64, Error> {
        match self.value(key)? {
            Value::Scalar(Scalar::Float(n)) => Ok(*n),
            Value::Scalar(Scalar::Unsigned(n)) => Ok(*n as f64),
            Value::Scalar(Scalar::Signed(n)) => Ok(*n as f64),
            _ => Err(invalid(format!("{}{key} must be numeric", self.prefix))),
        }
    }
}
pub fn inspect(directory: &Directory, identity: ArtifactIdentity) -> Result<Description, Error> {
    let architecture = directory
        .value("general.architecture")
        .and_then(Value::string)
        .ok_or_else(|| invalid("missing GGUF architecture"))?;
    let routed = match architecture {
        "qwen35" => false,
        "qwen35moe" => true,
        _ => {
            return Err(invalid(format!(
                "unsupported Qwen GGUF architecture {architecture:?}"
            )))
        }
    };
    let m = Metadata {
        directory,
        prefix: format!("{architecture}."),
    };
    if m.optional("rope.scaling.type")
        .is_some_and(|v| v.string() != Some("none"))
    {
        return Err(invalid(
            "scaled Qwen rotary definitions are not yet qualified",
        ));
    }
    let block_count = m.integer("block_count")?;
    let speculative = match m.optional("nextn_predict_layers") {
        None => 0,
        Some(v) => v
            .unsigned()
            .filter(|n| *n < block_count)
            .ok_or_else(|| invalid("nextn_predict_layers must be below block count"))?,
    };
    let layer_count = block_count - speculative;
    // Each real layer requires distinct stored weights. Reject impossible counts
    // before allocating architecture vectors from untrusted metadata.
    if layer_count > directory.tensors.len() as u64 {
        return Err(invalid("Qwen layer count exceeds available weight roles"));
    }
    let layers = match m.optional("attention.recurrent_layers") {
        None => {
            let interval = m.integer("full_attention_interval")?;
            (0..layer_count)
                .map(|i| {
                    if (i + 1) % interval == 0 {
                        MixerKind::Attention
                    } else {
                        MixerKind::Recurrent
                    }
                })
                .collect::<Vec<_>>()
        }
        Some(Value::Array(flags)) if flags.len() as u64 == layer_count => flags
            .iter()
            .map(|v| match v {
                Scalar::Bool(true) => Ok(MixerKind::Recurrent),
                Scalar::Bool(false) => Ok(MixerKind::Attention),
                _ => Err(invalid("invalid per-layer recurrent flags")),
            })
            .collect::<Result<Vec<_>, _>>()?,
        _ => return Err(invalid("invalid per-layer recurrent flags")),
    };
    let embedding = directory
        .tensor("token_embd.weight")
        .ok_or_else(|| invalid("missing Qwen embedding"))?;
    if embedding.shape.len() != 2 {
        return Err(invalid("Qwen embedding must be a matrix"));
    }
    if let Some(v) = directory.value("tokenizer.ggml.tokens") {
        if !matches!(v,Value::Array(a) if a.len() as u64==embedding.shape[0]) {
            return Err(invalid("tokenizer vocabulary differs from embedding table"));
        }
    }
    let Value::Array(sections) = m.value("rope.dimension_sections")? else {
        return Err(invalid("Qwen rotary sections must contain four integers"));
    };
    let sections = sections
        .iter()
        .map(|v| match v {
            Scalar::Unsigned(n) => Some(*n),
            Scalar::Signed(n) => u64::try_from(*n).ok(),
            _ => None,
        })
        .collect::<Option<Vec<_>>>()
        .and_then(|v| <[u64; 4]>::try_from(v).ok())
        .ok_or_else(|| invalid("Qwen rotary sections must contain four nonnegative integers"))?;
    let experts = if routed {
        Some(ExpertGeometry {
            count: m.integer("expert_count")?,
            selected: m.integer("expert_used_count")?,
            intermediate: m.integer("expert_feed_forward_length")?,
            shared_intermediate: m.integer("expert_shared_feed_forward_length")?,
            normalize_selected: true,
        })
    } else {
        None
    };
    let g = Geometry {
        activation_dtype: DType::BF16,
        hidden: m.integer("embedding_length")?,
        intermediate: if let Some(e) = &experts {
            e.intermediate
        } else {
            m.integer("feed_forward_length")?
        },
        vocabulary: embedding.shape[0],
        context_limit: m.integer("context_length")?,
        layers,
        attention_heads: m.integer("attention.head_count")?,
        kv_heads: m.integer("attention.head_count_kv")?,
        attention_width: m.integer("attention.key_length")?,
        rotary_width: m.integer("rope.dimension_count")?,
        rotary_base: m.number("rope.freq_base")?,
        rotary_sections: sections,
        epsilon: m.number("attention.layer_norm_rms_epsilon")?,
        convolution_width: m.integer("ssm.conv_kernel")?,
        recurrent_key_heads: m.integer("ssm.group_count")?,
        recurrent_value_heads: m.integer("ssm.time_step_rank")?,
        recurrent_width: m.integer("ssm.state_size")?,
        recurrent_head_mapping: HeadMapping::Tiled,
        experts,
    };
    g.validate()?;
    let recurrent_inner = product(&[g.recurrent_value_heads, g.recurrent_width])?;
    if m.integer("attention.value_length")? != g.attention_width
        || m.integer("ssm.inner_size")? != recurrent_inner
    {
        return Err(invalid(
            "Qwen attention/recurrent inner size differs from head geometry",
        ));
    }
    let mut consumed = HashSet::new();
    let mut tensor = |name: &str, shape: &[u64]| -> Result<WeightDescriptor, Error> {
        let value = directory
            .tensor(name)
            .ok_or_else(|| invalid(format!("missing Qwen weight {name:?}")))?;
        if value.shape != shape {
            return Err(invalid(format!(
                "Qwen weight {name:?}: expected {shape:?}, received {:?}",
                value.shape
            )));
        }
        product(shape)?;
        consumed.insert(name.to_string());
        Ok(WeightDescriptor {
            name: name.into(),
            shape: shape.into(),
            transform: Transform::Identity,
        })
    };
    let embedding = tensor(&embedding.name, &[g.vocabulary, g.hidden])?;
    let output_norm = tensor("output_norm.weight", &[g.hidden])?;
    let output = tensor(
        if directory.tensor("output.weight").is_some() {
            "output.weight"
        } else {
            &embedding.name
        },
        &[g.vocabulary, g.hidden],
    )?;
    let mut blocks = Vec::new();
    for (i, kind) in g.layers.iter().enumerate() {
        let p = format!("blk.{i}.");
        let mut weight = |name: &str, shape: &[u64]| tensor(&format!("{p}{name}"), shape);
        let mixer = if *kind == MixerKind::Attention {
            MixerWeights::Attention(Box::new(AttentionWeights {
                query_gate: weight(
                    "attn_q.weight",
                    &[
                        product(&[2, g.attention_heads, g.attention_width])?,
                        g.hidden,
                    ],
                )?,
                key: weight(
                    "attn_k.weight",
                    &[product(&[g.kv_heads, g.attention_width])?, g.hidden],
                )?,
                value: weight(
                    "attn_v.weight",
                    &[product(&[g.kv_heads, g.attention_width])?, g.hidden],
                )?,
                query_norm: weight("attn_q_norm.weight", &[g.attention_width])?,
                key_norm: weight("attn_k_norm.weight", &[g.attention_width])?,
                output: weight(
                    "attn_output.weight",
                    &[g.hidden, product(&[g.attention_heads, g.attention_width])?],
                )?,
            }))
        } else {
            MixerWeights::Recurrent(Box::new(RecurrentWeights {
                query_key_value: weight("attn_qkv.weight", &[g.recurrent_channels()?, g.hidden])?,
                gate: weight("attn_gate.weight", &[recurrent_inner, g.hidden])?,
                alpha: weight("ssm_alpha.weight", &[g.recurrent_value_heads, g.hidden])?,
                beta: weight("ssm_beta.weight", &[g.recurrent_value_heads, g.hidden])?,
                convolution: weight(
                    "ssm_conv1d.weight",
                    &[g.recurrent_channels()?, g.convolution_width],
                )?,
                decay: weight("ssm_a", &[g.recurrent_value_heads])?,
                time_bias: weight("ssm_dt.bias", &[g.recurrent_value_heads])?,
                norm: weight("ssm_norm.weight", &[g.recurrent_width])?,
                output: weight("ssm_out.weight", &[g.hidden, recurrent_inner])?,
            }))
        };
        let feedforward = if let Some(e) = &g.experts {
            FeedForwardWeights::Routed(Box::new(RoutedFeedForwardWeights {
                router: weight("ffn_gate_inp.weight", &[e.count, g.hidden])?,
                shared_router: weight("ffn_gate_inp_shexp.weight", &[g.hidden])?,
                expert_gate: weight("ffn_gate_exps.weight", &[e.count, e.intermediate, g.hidden])?,
                expert_up: weight("ffn_up_exps.weight", &[e.count, e.intermediate, g.hidden])?,
                expert_down: weight("ffn_down_exps.weight", &[e.count, g.hidden, e.intermediate])?,
                shared_gate: weight("ffn_gate_shexp.weight", &[e.shared_intermediate, g.hidden])?,
                shared_up: weight("ffn_up_shexp.weight", &[e.shared_intermediate, g.hidden])?,
                shared_down: weight("ffn_down_shexp.weight", &[g.hidden, e.shared_intermediate])?,
            }))
        } else {
            FeedForwardWeights::Dense(Box::new(DenseFeedForwardWeights {
                gate: weight("ffn_gate.weight", &[g.intermediate, g.hidden])?,
                up: weight("ffn_up.weight", &[g.intermediate, g.hidden])?,
                down: weight("ffn_down.weight", &[g.hidden, g.intermediate])?,
            }))
        };
        blocks.push(BlockWeights {
            input_norm: weight("attn_norm.weight", &[g.hidden])?,
            mixer,
            feedforward_norm: weight("post_attention_norm.weight", &[g.hidden])?,
            feedforward,
        });
    }
    for t in &directory.tensors {
        if consumed.contains(&t.name) {
            continue;
        }
        let speculative_role = t
            .name
            .strip_prefix("blk.")
            .and_then(|s| s.split_once('.'))
            .and_then(|(i, _)| i.parse::<u64>().ok().filter(|n| i == n.to_string()))
            .is_some_and(|i| i >= layer_count && i < block_count);
        if !speculative_role {
            return Err(invalid(format!(
                "Qwen artifact contains unbound weight role {:?}",
                t.name
            )));
        }
    }
    Ok(Description {
        artifact_identity: identity,
        geometry: g,
        embedding,
        output_norm,
        output,
        blocks,
    })
}
